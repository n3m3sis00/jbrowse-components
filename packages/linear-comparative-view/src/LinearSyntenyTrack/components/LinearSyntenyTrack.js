import { observer, PropTypes as MobxPropTypes } from 'mobx-react'
import PropTypes from 'prop-types'
import React from 'react'
import LinearSyntenyConnections from './LinearSyntenyConnections'

function LinearSyntenyTrack(props) {
  const { track, children } = props
  return (
    <>
      {track.trackMessageComponent ? (
        <track.trackMessageComponent model={track} />
      ) : (
        <LinearSyntenyConnections {...props} />
      )}
      {children}
    </>
  )
}
LinearSyntenyTrack.propTypes = {
  model: MobxPropTypes.objectOrObservableObject.isRequired,
  track: MobxPropTypes.objectOrObservableObject.isRequired,
  children: PropTypes.node,
}

LinearSyntenyTrack.defaultProps = {
  children: null,
}
export default observer(LinearSyntenyTrack)
