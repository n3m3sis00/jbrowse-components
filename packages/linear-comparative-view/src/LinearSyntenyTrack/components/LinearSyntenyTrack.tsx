import { observer, PropTypes as MobxPropTypes } from 'mobx-react'
import { Instance } from 'mobx-state-tree'
import PropTypes from 'prop-types'
import React from 'react'
import { LinearSyntenyTrackStateModel } from '..'

const LinearSyntenyTrack: React.FC<{
  model: Instance<LinearSyntenyTrackStateModel>
  children?: React.ReactNode
}> = props => {
  const { model, children } = props
  const { syntenyBlocks } = model
  return (
    <>
      <syntenyBlocks.ReactComponent {...props} />
      {children}
    </>
  )
}
LinearSyntenyTrack.propTypes = {
  model: MobxPropTypes.objectOrObservableObject.isRequired,
  children: PropTypes.element,
}

LinearSyntenyTrack.defaultProps = {
  children: null,
}
export default observer(LinearSyntenyTrack)
