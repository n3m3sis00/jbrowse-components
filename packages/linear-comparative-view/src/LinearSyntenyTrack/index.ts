/* eslint-disable @typescript-eslint/no-explicit-any */
import { types, getParent, addDisposer, Instance } from 'mobx-state-tree'
import { autorun } from 'mobx'
import { toArray } from 'rxjs/operators'
import CompositeMap from '@gmod/jbrowse-core/util/compositeMap'
import { getAdapter } from '@gmod/jbrowse-core/util/dataAdapterCache'
import { getContainingView } from '@gmod/jbrowse-core/util'
import { Feature } from '@gmod/jbrowse-core/util/simpleFeature'
import {
  getConf,
  ConfigurationReference,
  ConfigurationSchema,
} from '@gmod/jbrowse-core/configuration'

import {
  configSchemaFactory as baseConfig,
  stateModelFactory as baseModel,
} from '../LinearComparativeTrack'
import LinearSyntenyTrackComponent from './components/LinearSyntenyTrack'

export function configSchemaFactory(pluginManager: any) {
  return ConfigurationSchema(
    'LinearSyntenyTrack',
    {
      viewType: 'LinearSyntenyView',
      mcscanAnchors: {
        type: 'fileLocation',
        defaultValue: { uri: '/path/to/mcscan.anchors' },
      },
      geneAdapter1: pluginManager.pluggableConfigSchemaType('adapter'),
      geneAdapter2: pluginManager.pluggableConfigSchemaType('adapter'),
    },
    {
      baseConfiguration: baseConfig(pluginManager),
      explicitlyTyped: true,
    },
  )
}

export function stateModelFactory(pluginManager: any, configSchema: any) {
  return types
    .compose(
      baseModel(pluginManager, configSchema),
      types.model('LinearSyntenyTrack', {
        type: types.literal('LinearSyntenyTrack'),
        configuration: ConfigurationReference(configSchema),
      }),
    )
    .volatile(self => ({
      ReactComponent: (LinearSyntenyTrackComponent as unknown) as React.FC,
    }))
    .views(self => ({
      get subtrackViews() {
        return this.subtracks.map(subtrack =>
          getContainingView(subtrack),
        ) as any[]
      },
      get subtracks(): any[] {
        const subtracks: any[] = []
        const parentView = getParent(self, 2)
        // @ts-ignore
        parentView.views.forEach(subview => {
          subview.tracks.forEach((subviewTrack: any) => {
            if (this.trackIds.includes(getConf(subviewTrack, 'trackId'))) {
              subtracks.push(subviewTrack)
            }
          })
        })
        return subtracks
      },

      get trackFeatures() {
        return new CompositeMap<string, Feature>(
          this.subtracks.map(t => t.features),
        )
      },

      get adapterConfig() {
        return {
          type: 'MCScanAnchorsAdapter',
          mcscanAnchorsLocation: getConf(self, 'mcscanAnchors'),
          geneAdapter1: getConf(self, 'geneAdapter1'),
          geneAdapter2: getConf(self, 'geneAdapter2'),
        }
      },

      //       get layoutMatches() {
      //         const features = this.trackFeatures
      //         const alreadySeen = new Set<number>()
      //         const featNameMap: { [key: string]: Feature } = {}

      //         if (!self.anchorsData || !self.anchors) {
      //           return []
      //         } // this finds candidate features that share the same name

      //         for (const feature of features.values()) {
      //           const n = self.anchors[feature.get('name')]
      //           featNameMap[feature.get('name')] = feature

      //           if (n !== undefined) {
      //             alreadySeen.add(n)
      //           }
      //         }

      //         const blocks = Array.from(alreadySeen)
      //         const ret = []

      //         for (const block of blocks) {
      //           const r = self.anchorsData[block]
      //           if (r) {
      //             const r1 = featNameMap[r.name1]
      //             const r2 = featNameMap[r.name2]
      //             if (!r1 || !r2) {
      //               // alt logic (r1 && r2) || !(r3 && r4)
      //               // eslint-disable-next-line no-continue
      //               continue
      //             }
      //             const s1 = r1.get('start')
      //             const e1 = r1.get('end')
      //             const s2 = r2.get('start')
      //             const e2 = r2.get('end')

      //             ret.push([
      //               {
      //                 layout: [s1, 0, e1, 10] as LayoutRecord,
      //                 feature: new SimpleFeature({
      //                   data: { uniqueId: `${block}-1`, start: s1, end: e1 },
      //                 }),
      //                 level: 1,
      //                 refName: r1.get('refName') as string,
      //               },
      //               {
      //                 layout: [s2, 0, e2, 10] as LayoutRecord,
      //                 feature: new SimpleFeature({
      //                   data: { uniqueId: `${block}-2`, start: s2, end: e2 },
      //                 }),
      //                 level: 0,
      //                 refName: r2.get('refName') as string,
      //               },
      //             ])
      //           }
      //         }
      //         return ret
      //       },

      get trackIds() {
        return getConf(self, 'trackIds') as string[]
      },
    }))
    .actions(self => ({
      afterAttach() {
        addDisposer(
          self,
          autorun(async () => {
            const adapter = getAdapter(
              pluginManager,
              getConf(self, 'trackId'),
              self.adapterConfig.type,
              self.adapterConfig,
              null,
              null,
            ).dataAdapter
            const r = getParent(self, 2) as any
            const blockFeats = r.views.map(subview => {
              return Promise.all(
                subview.staticBlocks.map(async block => {
                  if (block.refName !== undefined) {
                    const { refName, start, end, assemblyName } = block
                    const observable = adapter.getFeatures({
                      refName,
                      start,
                      end,
                      assemblyName,
                    })
                    return observable.pipe(toArray()).toPromise()
                  }
                  // might be an interregion padding block, return empty
                  return []
                }),
              )
            })
            console.log(Promise.all(blockFeats))
          }),
        )
      },
    }))
}

export type LinearSyntenyTrack = ReturnType<typeof stateModelFactory>
export type LinearSyntenyTrackModel = Instance<LinearSyntenyTrack>
