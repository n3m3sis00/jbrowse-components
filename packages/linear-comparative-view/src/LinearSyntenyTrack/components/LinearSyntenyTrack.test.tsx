// eslint-disable-next-line import/no-extraneous-dependencies
import { render } from '@testing-library/react'
import { types, getSnapshot } from 'mobx-state-tree'
import ViewType from '@gmod/jbrowse-core/pluggableElementTypes/ViewType'
import React from 'react'
import Plugin from '@gmod/jbrowse-core/Plugin'
import PluginManager from '@gmod/jbrowse-core/PluginManager'
import AdapterType from '@gmod/jbrowse-core/pluggableElementTypes/AdapterType'
// eslint-disable-next-line import/no-extraneous-dependencies
import {
  AdapterClass as NCListAdapter,
  configSchema as nclistAdapterConfigSchema,
} from '@gmod/jbrowse-plugin-jbrowse1/src/NCListAdapter'
import LinearSyntenyTrack from './LinearSyntenyTrack'
import {
  AdapterClass as MCScanAnchorsAdapter,
  configSchema as mcscanAdapterConfigSchema,
} from '../../MCScanAnchorsAdapter'
import { stateModelFactory, configSchemaFactory } from '..'

class MCScanAdapterPlugin extends Plugin {
  // @ts-ignore
  install(pluginManager: any) {
    pluginManager.addAdapterType(
      () =>
        new AdapterType({
          name: 'MCScanAnchorsAdapter',
          AdapterClass: MCScanAnchorsAdapter,
          configSchema: mcscanAdapterConfigSchema,
        }),
    )
  }
}

class NCListAdapterPlugin extends Plugin {
  // @ts-ignore
  install(pluginManager: any) {
    pluginManager.addAdapterType(
      () =>
        new AdapterType({
          name: 'NCListAdapter',
          AdapterClass: NCListAdapter,
          configSchema: nclistAdapterConfigSchema,
        }),
    )
  }
}

class FakeTrackAndViewPlugin extends Plugin {
  // @ts-ignore
  install(pluginManager: any) {
    const FakeTrack = types
      .model('FakeTrack', {
        trackId: 'FakeTrack',
        type: 'FakeTrack',
        features: types.frozen(),
        layoutFeatures: types.frozen(),
        configuration: types.frozen(),
      })
      .actions(self => ({
        afterCreate() {
          self.features = new Map(Object.entries(self.features))
          self.layoutFeatures = new Map(Object.entries(self.layoutFeatures))
        },
      }))
    pluginManager.addViewType(
      () =>
        new ViewType({
          name: 'LinearGenomeView',
          stateModel: types
            .model('LinearGenomeView', {
              type: 'LinearGenomeView',
              tracks: types.array(FakeTrack),
            })
            .views(self => ({
              getTrack(trackConfigId: string) {
                return self.tracks.find(
                  t => t.configuration.trackId === trackConfigId,
                )
              },
            })),
          ReactComponent,
          RenderingComponent: true,
        }),
    )
  }
}
const ReactComponent = () => <p>Hello World</p>
const getView = () => {
  const pluginManager = new PluginManager([
    new MCScanAdapterPlugin(),
    new NCListAdapterPlugin(),
    new FakeTrackAndViewPlugin(),
  ])

  pluginManager.configure()
  const configSchema = configSchemaFactory(pluginManager)
  return types.model('LinearSyntenyView', {
    type: 'LinearSyntenyView',
    views: types.array(types.frozen()),
    tracks: types.array(stateModelFactory(pluginManager, configSchema)),
  })
}
const createSyntenyTrack = () =>
  // @ts-ignore
  getView().create({
    tracks: [
      {
        configuration: {
          trackId: 'trackId0',
          name: 'synteny',
          type: 'LinearSyntenyTrack',
          adapter: {
            type: 'MCScanAnchorsAdapter',
            assemblyNames: ['peach', 'grape'],
            mcscanAnchorsLocation: {
              uri: 'test_data/grape.peach.anchors',
            },
            subadapters: [
              {
                rootUrlTemplate:
                  'https://s3.amazonaws.com/jbrowse.org/genomes/synteny/peach_gene/{refseq}/trackData.json',
              },
              {
                rootUrlTemplate:
                  'https://s3.amazonaws.com/jbrowse.org/genomes/synteny/grape_gene/{refseq}/trackData.json',
              },
            ],
          },
        },
        renderDelay: 100,
        syntenyBlocks: { key: 'test' },
        type: 'LinearSyntenyTrack',
      },
    ],
  })

// function timeout(ms) {
//   return new Promise(resolve => setTimeout(resolve, ms))
// }
test('test rendering synteny data', async () => {
  const view = createSyntenyTrack()

  const { findByTestId, container } = render(
    <LinearSyntenyTrack model={view.tracks[0]} />,
  )
  await findByTestId('loading-synteny')

  expect(container).toMatchSnapshot()
  expect(1).toBe(1)
})
