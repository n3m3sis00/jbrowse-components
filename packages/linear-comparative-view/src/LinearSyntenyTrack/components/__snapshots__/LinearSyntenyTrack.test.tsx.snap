// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test rendering synteny data 1`] = `
<div>
  <div
    class="makeStyles-loading"
    data-testid="loading-synteny"
  >
    Loading â€¦
    <div
      class="MuiLinearProgress-root MuiLinearProgress-colorPrimary MuiLinearProgress-indeterminate"
      role="progressbar"
    >
      <div
        class="MuiLinearProgress-bar MuiLinearProgress-barColorPrimary MuiLinearProgress-bar1Indeterminate"
      />
      <div
        class="MuiLinearProgress-bar MuiLinearProgress-bar2Indeterminate MuiLinearProgress-barColorPrimary"
      />
    </div>
  </div>
</div>
`;
